### PR TITLE
fix: don't call binding converter with null values unless only the final segment of the path is null

### LINKED ISSUE #16016: Converter receiving unexpected `null` value
### Current behavior

Converter is receiving `null` value when it should not be possible.

### Expected behavior

No `null` value

### How to reproduce it (as minimally and precisely as possible)

[uno-bug-repro-converter-null-main.zip](https://github.com/unoplatform/uno/files/14742677/uno-bug-repro-converter-null-main.zip)

### Workaround

Gracefully handle `null` values in converters.

### Works on UWP/WinUI

Yes 

### Environment

_No response_

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by ramezgerges**
GitHub Issue (If applicable): closes #16016

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Review by Youssef1313 — COMMENTED** [review]
It's better to write the tests using XAML file so that we can get them validated on WinUI as well.

Also, good to test FallbackValue behavior, and test for both `Binding` and `x:Bind`

**Comment by jeromelaban** [comment]
@ramezgerges could you take a look? It seems that some tests are not passing.

File: src/Uno.UI.RuntimeTests/IntegrationTests/common/TestServices.KeyboardHelper.cs [hunk 1]
```diff
@@ -318,6 +319,18 @@ public static void GamepadDpadDown(UIElement element = null)
 			{
 				PressKeySequence("$d$_GamepadDpadDown#$u$_GamepadDpadDown", element);
 			}
+
+			/// <param name="text">Assuming lowercase text. To add capitalization, use <see cref="PressKeySequence"/></param>
+			public static void InputText(string text, UIElement element = null)
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/IntegrationTests/common/TestServices.KeyboardHelper.cs:324]: any explanation on why this was added?
from what ive seen, this doesnt work: no visual change & no event raised

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs [hunk 2]
```diff
@@ -280,47 +284,82 @@ public async Task When_Text_Changed_SuggestionChosen()
...
-					AutoSuggestionBoxTextChangeReason.UserInput,
-					AutoSuggestionBoxTextChangeReason.ProgrammaticChange,
-					AutoSuggestionBoxTextChangeReason.SuggestionChosen
-				},
-				reasons);
+			await WindowHelper.WaitForIdle();
+
+#if __SKIA__
+			// skia is closer to what happens on WinUI. On WinUI, if there is no delay between changes, 
+			// AutoSuggestBox.TextChanged is fired once (but TextBox.TextChanged fires everytime)
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs:344]: @ramezgerges 
is this even intended? i assuming all of uno code would behave similar, unless underlying platform or platform-specific code suggest otherwise
this feel like a bug that should be fixed, or [Ignore]'d rather than getting a special passing setup...

Comment (ramezgerges) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs:344]: The intended behaviour is to match WinUI (I don't think the WinUI devs would classify this as a bug). The behviour in Uno is not platform-agnostic due to the way TextBox works, which differs between the managed (skia) and native (mobile) implementations. I don't remember very well, but I think AutoSuggestBox.TextChanges has a real reason for behaving like this. Conceptually, I think AutoSuggestBox only cares about entire "text runs", not individual character changes. This reduces the amount of suggestion updates that need to happen (e.g. like how server-side input-validation doesn't happen on each keystroke).

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs [hunk 3]
```diff
@@ -339,11 +347,11 @@ public async Task When_Text_Changed_Sequence(bool waitBetweenActions)
...
-			// skia is closer to what happens on WinUI. On WinUI, if there is no delay between changes, 
-			// AutoSuggestBox.TextChanged is fired once (but TextBox.TextChanged fires everytime)
+#if __SKIA__ // We want to test the behaviour of "typing individual characters in sequence", not setting the Text in one shot. The behaviour is currently only accurate on skia.
 			if (!waitBetweenActions)
 			{
+				// skia is closer to what happens on WinUI. On WinUI, if there is no delay between changes,
+				// AutoSuggestBox.TextChanged is fired once (but TextBox.TextChanged fires everytime)
 				expectations = new() { SuggestionChosen };
 			}
 #endif
```
Comment (Xiaoy312) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_AutoSuggestBox.cs:357]: I just took a look at the test results, it seems that on ios,droid,wasm, we only get a single event happening:
```
Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: 
CollectionAssert.AreEquivalent failed. The number of elements in the collections do not match. Expected:<7>. Actual:<1>.
expectations[7]: SuggestionChosen,ProgrammaticChange,UserInput,SuggestionChosen,UserInput,ProgrammaticChange,SuggestionChosen; 
actual[1]: SuggestionChosen
```
so we can remove the __SKIA__, since the behavior can be expected everywhere else.

File: src/Uno.UI/UI/Xaml/Data/Binding.cs [hunk 4]
```diff
@@ -217,6 +216,9 @@ public object CompiledSource
 		/// </summary>
 		internal string[] XBindPropertyPaths { get; private set; }
 
+		// Each of these values could be null and the Binding could still be an x:Bind, but they can't all be null
+		public bool IsXBind => XBindSelector is not null || XBindPropertyPaths is not null || CompiledSource is not null || XBindBack is not null;
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Data/Binding.cs:220]: This shouldn't be public

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Data/Binding.cs:220]: Let's make it an extension on markuphelper

File: src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.cs [hunk 5]
```diff
@@ -623,36 +623,27 @@ protected virtual void OnFontStyleChanged(FontStyle oldValue, FontStyle newValue
 
 	protected virtual void OnContentChanged(object oldValue, object newValue)
 	{
-		if (oldValue != null && newValue == null)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.cs:626]: Why did you remove this part?

File: src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.cs [hunk 6]
```diff
@@ -623,36 +623,27 @@ protected virtual void OnFontStyleChanged(FontStyle oldValue, FontStyle newValue
...
 
-			TrySetDataContextFromContent(newValue);
+		TrySetDataContextFromContent(newValue);
 
-			SetUpdateTemplate();
-		}
-		else
-		{
-			// Restore the inherited data context as it may have been overridden by TrySetDataContextFromContent
-			this.ClearValue(DataContextProperty, DependencyPropertyValuePrecedences.Local);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.cs:649]: Same here, why did you remove it?

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Data/Given_BindingExpression.cs [hunk 7]
```diff
@@ -0,0 +1,585 @@
+using System.Threading.Tasks;
+using Microsoft.UI.Xaml.Controls;
+using Microsoft.UI.Xaml.Media;
+using Private.Infrastructure;
+using Uno.Disposables;
+namespace Uno.UI.RuntimeTests.Tests.Windows_UI_Xaml_Data;
+
+[TestClass]
+public class Given_BindingExpression
```
Comment (jeromelaban) [inline src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Data/Given_BindingExpression.cs:9]: Can you validate that those new tests pass on the winui version of the samples app?

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/TextBoxTests/TextBoxTests.cs [hunk 8]
```diff
@@ -323,7 +323,7 @@ public void TextBox_Formatting_FlickerText()
 			using (new AssertionScope())
 			{
 				text2.Should().Be(text1, because: "Text content should not change at max length.");
-				text3.Should().Be("TextChanged: 1");
+				text3.Should().Be("TextChanged: 2");
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/TextBoxTests/TextBoxTests.cs:326]: What's different in this context?

Comment (ramezgerges) [reply src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/TextBoxTests/TextBoxTests.cs:326]: During a "typing sequence" we previously only fired TextChanged once at the end. To match WinUI (and to fix an autosuggestbox-related problem), we now fire it each time. In this test, the text is set but then set again by the IInputFilter, so previously a single TextChanged was fired at the end, now it gets fired twice.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Data/Given_BindingExpression.cs (hunk 7 above)
Comment (ramezgerges) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Data/Given_BindingExpression.cs:9]: I wrote them on WinUI first then made our implementation pass, so they should definitely pass.

File: src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.cs (hunk 6 above)
Comment (ramezgerges) [reply src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.cs:649]: Same as above (don't clear the DataContext, just set it to null like any other value).

**Review by jeromelaban — APPROVED** [review]
